### PR TITLE
Issue 1373: Close connection after getting info

### LINKED ISSUE #1373: Connection leak in SegmentStreamFactoryImpl
**Problem description**
SegmentStreamFactoryImpl opens a connection to get the transaction info, but this connection is leaked.

**Problem location**
SegmentStreamFactoryImpl

**Suggestions for an improvement**
Close the connection in a finally block

### PR DESCRIPTION
**Pull request by tkaitchuck**
**Change log description**
* Prevent a connection leak in SegmentOutputStreamFactory
* Add some tests for the client.

**Purpose of the change**
Fixes #1373 

**What the code does**
Calls close on the connection that is used to get info on the transaction.
Adds tests covering error handling in SegmentOutputStream.

**How to verify it**
All tests should pass including new ones.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/netty/impl/ClientConnectionInboundHandler.java [hunk 1]
```diff
@@ -66,7 +66,7 @@ public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
     }
 
     @Override
-    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
```
Comment (shrids) [inline client/src/main/java/io/pravega/client/netty/impl/ClientConnectionInboundHandler.java]: +1
`channelInactive ()` this is invoked when the channel registered is now inactive and reached its end of lifetime.
Also, `channelUnregistered()` is always invoked when `channelInactive()` is invoked.

File: client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamFactoryImpl.java [hunk 2]
```diff
@@ -63,19 +62,24 @@ public void processingFailure(Exception error) {
...
                 connection.send(new WireCommands.GetTransactionInfo(1, segment.getScopedName(), txId));
             } catch (ConnectionFailedException e) {
                 throw new RuntimeException(e);
-            } 
-        }).exceptionally((Throwable t) -> {
-            name.completeExceptionally(t);
-            return null;
+            }
+        }).whenComplete((v, t) -> {
+            if (t!=null) {
```
Comment (shrids) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamFactoryImpl.java]: nit: checkstyle issue will cause a build failure.

File: controller/src/main/java/io/pravega/controller/server/ControllerService.java [hunk 3]
```diff
@@ -282,7 +282,7 @@ private SegmentRange convert(final String scope,
         return streamTransactionMetadataTasks.commitTxn(scope, stream, txId, null)
                 .handle((ok, ex) -> {
                     if (ex != null) {
-                        // TODO: return appropriate failures to user.
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/server/ControllerService.java]: I'm not sure what the intent of this TODO was, but perhaps it was referring to being more specific about the failure. In any case, I want to double-check why we are removing this TODO here and the other one below.

File: client/src/test/java/io/pravega/client/segment/impl/SegmentOutputStreamTest.java [hunk 4]
```diff
@@ -213,21 +207,98 @@ public void testFlush() throws ConnectionFailedException, SegmentSealedException
...
+        assertEquals(false, acked.isDone());
+        inOrder.verifyNoMoreInteractions();
     }
 
+    
     @Test
-    @Ignore
-    public void testOutOfOrderAcks() {
-        fail();
+    public void testFailOnClose() throws ConnectionFailedException, SegmentSealedException {
```
Comment (fpj) [inline client/src/test/java/io/pravega/client/segment/impl/SegmentOutputStreamTest.java]: This test case seems to be doing the right thing, but I find it hard to read. It is not strictly necessary, but adding comments that help guide the code reader, makes it easier to determine whether the behavior is correct. I suppose the other argument could be made to, that not adding any comment forces the code reviewer to understand the code and verify that it works. There is probably some middle ground....

File: client/src/main/java/io/pravega/client/netty/impl/ClientConnectionInboundHandler.java (hunk 1 above)
Comment (fpj) [inline client/src/main/java/io/pravega/client/netty/impl/ClientConnectionInboundHandler.java]: I don't think there is documentation explaining this well, could we please add a comment block to explain how these channel methods are called and what we expect to happen?

File: controller/src/main/java/io/pravega/controller/server/ControllerService.java (hunk 3 above)
Comment (tkaitchuck) [inline controller/src/main/java/io/pravega/controller/server/ControllerService.java]: Reverted. I also opened #1401 and #1402 with more specific information.

File: client/src/main/java/io/pravega/client/netty/impl/ClientConnectionInboundHandler.java (hunk 1 above)
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/netty/impl/ClientConnectionInboundHandler.java]: I added a javadoc, with a pointer to the actual javadoc, which for whatever reason netty did not include on the parent class.

File: client/src/test/java/io/pravega/client/segment/impl/SegmentOutputStreamTest.java (hunk 4 above)
Comment (tkaitchuck) [inline client/src/test/java/io/pravega/client/segment/impl/SegmentOutputStreamTest.java]: Added

**Pull request merged**